build/env.py

```diff
@@ -0,0 +1,98 @@
+import os
+import shutil
+import subprocess
+import sys
+import sysconfig
+import tempfile
+import types
+
+
+if False:  # TYPE_CHECKING  # pragma: no cover
+    from typing import Dict, Optional, Iterable, Type
+
+
+class IsolatedEnvironment(object):
+    '''
+    Isolated build environment context manager
+
+    Non-standard paths injected directly to sys.path still be passed to the environment.
+    '''
+
+    def __init__(self):  # type: () -> None
+        self._env = {}  # type: Dict[str, Optional[str]]
+
+    def _replace_env(self, key, new):  # type: (str, Optional[str]) -> None
+        if not new:  # pragma: no cover
+            return
+
+        self._env[key] = os.environ.get(key, None)
+        os.environ[key] = new
+
+    def _restore_env(self):  # type: () -> None
+        for key, val in self._env.items():
+            if val is None:
+                os.environ.pop(key, None)
+            else:
+                os.environ[key] = val
+
+    def _get_env_path(self, path):  # type: (str) -> Optional[str]
+        return sysconfig.get_path(path, vars=self._env_vars)
+
+    def __enter__(self):  # type: () -> IsolatedEnvironment
+        self._path = tempfile.mkdtemp(prefix='build-env-')
+        self._env_vars = {
+            'base': self._path,
+            'platbase': self._path,
+        }
+
+        sys_path = sys.path[1:]
+
+        remove_paths = os.environ.get('PYTHONPATH', '').split(os.pathsep)
+
+        if os.name == 'nt':  # pragma: no cover
```

Comment:
do we need to define these? can't we acquire it from sysconfig?

Reply:
Yes, but I don't want to add a test that is just same the logic.

Reply:
At the very least should be global. What I was saying though is that will OS always contain this? pypy for example defines its own schemes and any python distribution has a right to define so. IMHO we should always use sysconfig one.

Reply:
> At the very least should be global. What I was saying though is that will OS always contain this? pypy for example defines its own schemes and any python distribution has a right to define so.

I do not follow.

Reply:
I gather you took this list from https://docs.python.org/3/library/sysconfig.html#installation-paths but that's only true for CPython as distributed by the CPython core developer team. The scheme list may be extended and often is by other distributions (such as Debian repackaging), or other python implementations, see here for pypy https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/lib-python/2.7/sysconfig.py#L29-38 

So you should use https://docs.python.org/3/library/sysconfig.html#sysconfig.get_scheme_names

Reply:
Ah, yes. Github doesn't show the full code and I thought this was other line :joy:.